### PR TITLE
Added the ability to provide optional documentation for each command, to be displayed in the help output.

### PR DESCRIPTION
**Pull request by brucemack**
Goal is to improve ease-of-use for applications with complex command sets.  Per advice from Phil, doing this in a way that does't add RAM overhead.

### CODE REVIEW

File: extras/tests/HelpTest/SimulatedStream.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+#include <Arduino.h>
+#include "../simpleSerialShellTest/simpleFIFO.h"
+
+/**
+ * @brief A simple implementation fo the Stream interface.  Useful for 
+ * creating mock Serial objects for unit testing.
+ * 
+ * A header-only class for ease of re-use across tests cases.
+ */
+template<int MAX_BUFSIZE>
```
Comment (philj404) [inline extras/tests/HelpTest/SimulatedStream.h:13]: First time I've had a template class in my own code base.  I guess I had to learn sometime....

File: src/SimpleSerialShell.cpp [hunk 2]
```diff
@@ -21,30 +25,69 @@ SimpleSerialShell::Command * SimpleSerialShell::firstCommand = NULL;
...
+            // Look for the documentation delimiter and make sure we don't 
+            // consider anything beyond it in the comparison.  
+            //
+            // Note for future consideration: The temporary String here could
+            // be eliminated here by leveraging strlen_P, pgm_read_byte, and 
+            // strncasecmp_P.  That will take a bit of research since 
+            // the header file may have a different name on ESP2886/ESP32.
+            String work(nameAndDocs);
+            int compareLength = SIMPLE_SERIAL_SHELL_BUFSIZE;
+            for (unsigned int i = 0; i < work.length(); i++) {
```
Comment (philj404) [inline src/SimpleSerialShell.cpp:53]: FYI an alternative could be to use String's built-in search: `work.indexOf(DOC_DELIMITER)` will find the delimiter for you, or return -1.

That said, this is still easy to understand.

**Comment by philj404** [comment]
I'm working through this pull request.  It seems straightforward.  I want to do a little local testing before I work it through the system.

Should the next release be 1.0?  I haven't had significant changes for a while.

**Comment by brucemack** [comment]
Sounds great.  I'll leave the release numbering decision to you.  Given how much functionality you have, it already feels like v4.5.1 to me. :-) 

For what it's worth, I don't use the Arduino IDE and I take most libraries from their main branch on Github.  So the formality of a release doesn't make any difference to me.  What matters is that the changes are acceptable and get merged to main.

**Comment by philj404** [comment]
@brucemack I have put these changes on the `test` branch.
In addition I modified the EchoCommand, ArduinoTextInterface and AdjustableBlink examples to add help hints.

Does this work for you?

**Comment by brucemack** [comment]
Yes, that's great.  I will switch over to using your test branch until this makes it up to main.  Your improved examples look good.  Thank you for considering this change.

**Comment by philj404** [comment]
I'm going to try a squash-and-merge of this PR to the `next` branch. I've tested it and it seems to work, but I made some additional changes.  I still want to show this was merged, rather than just closed.

`next` will get squashed-and-merged to `main` Real Soon Now, as soon as I get a few other issues straightened out.

**Comment by philj404** [comment]
oops.  Meant to change to `next`.  Oh well.

**Comment by philj404** [comment]
Your changes should be visible in the new 0.9.0 release.

**Pull request merged**